packages/api-plugin-carts/src/util/updateCartFulfillmentGroups.js

```diff
@@ -1,61 +1,123 @@
 import Random from "@reactioncommerce/random";
+import ReactionError from "@reactioncommerce/reaction-error";
 
 /**
- * @summary Figures out which fulfillment group a cart item should initially be in
+ * @summary Figures out from which fulfillment groups the item should be removed
  * @param {Object[]} currentGroups The current cart fulfillment groups array
- * @param {String[]} supportedFulfillmentTypes Array of fulfillment types supported by the item
+ * @param {String} itemId Id of the item to be removed
+ * @param {String} shopId The ID of the shop that owns the item (product)
+ * @returns {Object[]|null} The groups array or null if no viable group
+ */
+function determineRemoveFromGroups(currentGroups, itemId, shopId) {
+  let removeFromGroups = currentGroups.map((group) => {
+    if (group.itemIds && group.itemIds.includes(itemId) && shopId === group.shopId) {
+      return group;
+    }
+    return null;
+  });
+  removeFromGroups = removeFromGroups.filter((group) => !!group); // remove nulls
+  // If the item is present in more that one Fulfillment group, we remove it from all(for backward compatibility)
+  return removeFromGroups || null;
+}
+
+/**
+ * @summary Figures out which fulfillment group a cart item should be added to
+ * @param {Object[]} currentGroups The current cart fulfillment groups array
+ * @param {String[]} selectedFulfillmentType Selected Fulfillment type for the item
  * @param {String} shopId The ID of the shop that owns the item (product)
  * @returns {Object|null} The group or null if no viable group
  */
-function determineInitialGroupForItem(currentGroups, supportedFulfillmentTypes, shopId) {
-  const compatibleGroup = currentGroups.find((group) => supportedFulfillmentTypes.includes(group.type) &&
+function determineAddToGroup(currentGroups, selectedFulfillmentType, shopId) {
+  const compatibleGroup = currentGroups.find((group) => selectedFulfillmentType === group.type &&
     shopId === group.shopId);
   return compatibleGroup || null;
 }
 
+/**
+ * @summary Check if the provided fulfillment type is present in any of the groups and adds if not
+ * @param {Object[]} currentGroups The current cart fulfillment groups array
+ * @param {String} fulfillmentType Specific fulfillment type to be checked
+ * @param {Object} item The item (product) object
+ * @returns {undefined}
+ */
+function checkAndAddToGroup(currentGroups, fulfillmentType, item) {
+  let removeFromGroups = determineRemoveFromGroups(currentGroups, item._id, item.shopId);
+  if (removeFromGroups && removeFromGroups.length > 0) {
+    removeFromGroups = (removeFromGroups || []).map((group) => {
+      group.itemIds = (group.itemIds || []).filter((itemId) => item._id !== itemId);
+      return group;
+    });
+  }
```

Comment:
I think there are many redundant null checks.
There is no need to reassign `removeFromGroups`, for each is enough.
```suggestion
  const removeFromGroups = determineRemoveFromGroups(currentGroups, item._id, item.shopId);
removeFromGroups.forEach((group) => {
  group.itemIds = _.pull(group.itemIds, "itemId");
});
```